### ISSUE TITLE
Three-Valued Logic and the World of Issomething

### ISSUE THREAD

**Issue by dalum** [JuliaLang/julia#29665]
I couldn't find an issue discussing this, so now I'm opening one.

Right now, `iszero(missing)` returns `missing`.  But there is a lot of code in base (at least in `LinearAlgebra`) that relies on the assumption that `iszero` will return a Boolean value.  For instance, matrix inversion will do such a check at several stages, and error with a `SingularException` if `iszero` returns true [1].

In most of the uses I have seen, `iszero(x)` is used to branch or short–circuit a function in the case where `x` is zero *with certainty*, i. e., `iszero(x) === true`.  If `x` does not have this property, then the generic algorithm can proceed.  For instance, I think most would assume that an innocuous looking generic function like: `strongdiv(a, x) = iszero(a) && iszero(x) ? 0. : a/x`, should just work, as long as `zero(a)` is a well–defined concept.

At present, however, it won't.  It will fail when given a `missing`, although `a/missing` is a perfectly acceptable thing to do.

My own use case, which prompted me to consider this, involves a type that wraps a symbol, expression or number.  I. e., a symbolic type for doing symbolic manipulations.  In this case, a `Symbolic(:a)` is not known to be `zero(Symbolic)`, and so a symbolic `a == 0` just returns a wrapped expression object.  Right now, the default fallback in Base is `iszero(x) = x == zero(x)`, which means that `iszero(a)` would return the wrapped expression object, `a == 0`, and not a Boolean value.  Thus in order for my user–defined type to support the generic pattern above, `iszero(::Symbolic) = false` must be implemented.

This is of course not a major issue, it is probably only a few lines of code in most cases, but it raises the question: if almost all code assumes that `iszero` returns a `Bool`, should the default fallback return a `Bool`?  And should this be enforced through documentation?

The argument above extends to the whole issiverse of `isone`, `isinf`, `isfinite`, `isodd`, `iseven`, `isreal`, `isnan`, and any `is`-function that I've missed.

My proposal is that they should, i. e., that `iszero(x) = (x == zero(x)) === true`, and that the methods for `missing` should be updated accordingly.  This would make a clear distinction between `==` comparison and `is`-functions, and give the latter a more reliable behaviour in logical contexts.  The odd-one-out here is of course `isless`, which treats `missing` as larger than any other value, but that can be seen as an exception to a more general rule about certainty to preserve its use in ordering.

TL:DR; I propose that it should be the case that most functions starting with `is` reliably return a `Bool`, while `==` comparisons are allowed to return other objects, such as `missing`.

[1] (For a matrix with `missing` in it, the generic algorithm for matrix inverses with `missing` values presently fails much earlier because it wrongly tries to promote to `Float64`, but that is a tangential issue.)

**Comment by StefanKarpinski** [comment]
This is the problem with giving people any amount of logic for a missing value... everyone wants it to do something different in a different context. That said, it seems fairly wrong to have `iszero(missing)` be true. After all, `missing` is not really a zero value, is it?

**Comment by dalum** [comment]
The suggestion I made above would make `iszero(missing)` be false.  I'm sorry if my example made it look otherwise.  I'm not actually super concerned with the `missing` value in particular, I am only using it as an example because it illustrates the point a bit more clearly.  The problem is more broadly about having some reasonable guidelines for doing logic in the absence of certainty, which is unavoidable whether `missing` is there or not.

If it wasn't the case that `missing == missing` returns `missing`, I would probably have argued that the documentation should have stated that generic code assumes `==` returns a Boolean value.  But it seems like things have settled on `isequal` being an equality check in this sense.  If `zero(missing)` didn't return `missing`, which means `isequal(x, zero(x))` would return `true` for `missing`, this would probably also have been a fine candidate for generically checking for whether `x` is certainly `zero(x)`.

Alternatively, the least radical suggestion would be to have `iszero(x) === true` be the generic Boolean test, but, IMO, this is quite ugly and reminds me too much of JavaScript.

**Comment by StefanKarpinski** [comment]
But a `missing` value isn't definitely non-zero either...

**Comment by StefanKarpinski** [comment]
I'd rather that the code be non-generic and fail when passed a `missing` value than silently behave incorrectly. Things happen to work out in your `strongdiv` example, but that's certainly not guaranteed. Treating `missing` as definitely not zero is not valid three-valued logic.

**Comment by dalum** [comment]
I am not at all trying to argue that `missing` should be treated as non-zero—that would be a horrible idea.  What I am saying is that in `LinearAlgebra`, certain routines will check if an element or some temporary `iszero` and raise a `SingularException`, since then it knows that it would produce `NaN`s `Infs` or other non–sensical results.  However, if it is known that propagating a `missing` in this case does not produce a wrong result, but simply results in many of the entries to be "poisoned" by the `missing` value, then there is no reason not to let the algorithm proceed.  This should be a conscious choice made by the programmer and distinguished by what type of comparison is used.

In fact, all algorithms written in this generic fashion already allow `NaN`s to propagate, since `iszero(NaN)` is false.  One could equally well make the argument that `iszero(NaN)` is **not** definitely **not** zero, since `0/x` for sufficiently small `x` will at some point truncate to `0` and produce a `NaN`, although it would definitely have been `0` if the truncation didn't happen.  But since we cannot say with certainty that the `NaN` reflects a value that `iszero`, we let it pass.

**Comment by StefanKarpinski** [comment]
`NaN` is entirely different—it is a floating-point value distinct from every value on the real line: in particular, `NaN` is _definitely_ not zero.

**Comment by JeffBezanson** [comment]
It seems safer and clearer to me to use `iszero(x) === true` in cases that can tolerate missing, instead of making that behavior baked in to `iszero`.

**Comment by dalum** [comment]
Alright, I yield! Doing checks like `is...(x) === true` seems like the path of least resistance. :sweat_smile:

I do not think it should be the case that `isnan(missing) === false`, then.  A `NaN` value could be produced by an instrument as a piece of data.

**Comment by nalimilan** [comment]
Let's close this then.
